### ISSUE TITLE
bug: Catalog not updating for run

### ISSUE THREAD

**Issue by visch** [meltano/meltano#6763]
### Related topics:

- https://github.com/meltano/meltano/issues/2848

### Meltano Version

2.3.0 / 2.4.0 / 2.6.0

### Python Version

3.8

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

Windows 10, Ubuntu

### Description

Catalog isn't updating properly when using run. It's possible this is just with the `run` command and/or just with `Windows` I haven't verified anything beyond this setup. Repro steps below

### Code

```python
1. Get a postgres instance up my goto is `podman run -e POSTGRES_PASSWORD=postgres -p 5432:5432 -h postgres d postgres`
2. Create a new meltano project, `meltano init catalog_bug`
3. Add this `meltano.yml`

version: 1
default_environment: dev
plugins:
  extractors:
  - name: tap-postgres
    namespace: tap_postgres
    pip_url: git+https://github.com/AutoIDM/tap-postgres.git
    executable: tap-postgres
    capabilities:
    - state
    - catalog
    - discover
    settings:
    - name: sqlalchemy_url
      kind: password
    select:
    - autoidm-testh.col1
  loaders:
  - name: target-jsonl
    variant: andyh1203
    pip_url: target-jsonl
environments:
- name: dev
- name: staging
- name: prod
project_id: 0386f428-f486-4e8c-8ea3-75146356d2f1
```
4. `meltano install` (Note a `.env` var of `TAP_POSTGRES_SQLALCHEMY_URL="postgresql://postgres:postgres@localhost:5432/postgres"` maybe be helpful!)
5. Create some test data
```
CREATE TABLE autoidm.testh (
	col1 text NULL
);

insert into autoidm.testh (col1)
    VALUES('row1');
    
```
6. Run meltano `meltano run tap-postgres target-jsonl`
7. Alter the table schema
```
alter table autoidm.testh
add column col2 text NULL;

insert into autoidm.testh (col1, col2)
    VALUES('row2', 'row2');
```
8. Manually update `meltano.yml` to include the new column
```
    select:
    - autoidm-testh.col2
```
9. Run `meltano run tap-postgres target-jsonl` , note the new output data
```
{}
{}
``
Two empty rows as col2 doesn't exist
10. Run `rm -r .meltano\run\tap-postgres\` 
11. Run `meltano run tap-postgres target-jsonl` everything works now
```

**Comment by visch** [comment]
Verified this happens on Linux as well (Ubuntu atleast)

**Comment by tayloramurphy** [comment]
Thanks for this @visch - I added it to https://github.com/meltano/meltano/milestone/12 since we have quite a few catalog related issues. 

We're going to have to put more focused effort into catalog management in the coming months. cc @aaronsteers

**Comment by aaronsteers** [comment]
Compiling here a list of possible solutions that we have logged as of now:

- #6915 
- #6919 
- #2848 

And the general discussion thread for this topic:

- https://github.com/meltano/meltano/discussions/6757#discussioncomment-3650870

And a related issue for improving docs to make caching methodology clear:

- #6292

**Comment by edgarrmondragon** [comment]
Arguably addressed by #8580. Do comment if we're missing something and I'll reopen.
